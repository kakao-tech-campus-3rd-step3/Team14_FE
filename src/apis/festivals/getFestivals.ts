import { apiInstance } from '@/apis/instance';
import API_ENDPOINTS from '@/constants/apiEndpoints';
import type { Festival } from '@/types/FestivalType';
import type { ApiErrorResponse } from '@/apis/instance';
import type { AxiosResponse } from 'axios';
import { generatePath } from 'react-router-dom';

export interface GetFestivalsResponse {
  content: Festival[];
  pageable: {
    pageNumber: number;
    pageSize: number;
    sort: {
      empty: boolean;
      sorted: boolean;
      unsorted: boolean;
    };
    offset: number;
    paged: boolean;
    unpaged: boolean;
  };
  last: boolean;
  totalElements: number;
  totalPages: number;
  first: boolean;
  size: number;
  number: number;
  sort: {
    empty: boolean;
    sorted: boolean;
    unsorted: boolean;
  };
  numberOfElements: number;
}

export const getFestivals = async (params: {
  areaId: string;
}): Promise<AxiosResponse<GetFestivalsResponse, ApiErrorResponse>> => {
  return await apiInstance.get<GetFestivalsResponse>(
    generatePath(API_ENDPOINTS.FESTIVALS, { areaId: params.areaId }) + '?size=10000',
  );
};

export default getFestivals;
