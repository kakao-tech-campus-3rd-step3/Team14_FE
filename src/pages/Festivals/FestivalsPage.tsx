import Container from '@/components/common/Container';
import Footer from '@/components/common/Footer';
import Header from '@/components/common/Header';
import FestivalsAISection from '@/pages/Festivals/components/FestivalsAISection';
import FestivalsAreaSection from '@/pages/Festivals/components/FestivalsAreaSection';
import { Suspense } from 'react';
import { FestivalsSectionSkeleton } from '@/pages/Festivals/components/FestivalCardSkeleton';

const FestivalsPage = () => {
  return (
    <Container>
      <Header variant="page" />
      <div className="flex flex-col items-center px-8 py-4 gap-12">
        <Suspense fallback={<FestivalsSectionSkeleton count={3} />}>
          <FestivalsAISection />
        </Suspense>
        <Suspense fallback={<FestivalsSectionSkeleton count={3} />}>
          <FestivalsAreaSection />
        </Suspense>
      </div>
      <Footer initialSelected="none" />
    </Container>
  );
};

export default FestivalsPage;
