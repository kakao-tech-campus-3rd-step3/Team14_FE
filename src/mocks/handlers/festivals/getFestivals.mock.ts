import { apiBaseUrl } from '@/apis/instance';
import API_ENDPOINTS from '@/constants/apiEndpoints';
import { festivalsMockData } from '@/mocks/data/festivals.mock';
import { http, HttpResponse } from 'msw';

export const getFestivalsHandler = [
  http.get<{ areaId: string }>(apiBaseUrl + API_ENDPOINTS.FESTIVALS, async ({ params }) => {
    const { areaId } = params;

    if (!areaId || areaId === '0') {
      return HttpResponse.json(
        {
          status: 'BAD_REQUEST',
          statusCode: 400,
          message: '존재하지 않는 지역입니다',
        },
        { status: 400 },
      );
    }

    return HttpResponse.json({
      ...festivalsMockData,
    });
  }),
];
