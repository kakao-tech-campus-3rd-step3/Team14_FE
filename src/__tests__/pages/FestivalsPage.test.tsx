import { render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import FestivalsPage from '@/pages/Festivals/FestivalsPage';
import { festivalsMockData } from '@/mocks/data/festivals.mock';
import * as ReactRouter from 'react-router-dom';

vi.mock('react-router-dom', async () => ({
  ...(await vi.importActual('react-router-dom')),
}));

// QueryClient 설정
const createTestQueryClient = () =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        gcTime: 0,
      },
    },
  });

// 테스트용 래퍼 컴포넌트
const TestWrapper = ({ children }: { children: React.ReactNode }) => {
  const queryClient = createTestQueryClient();
  return (
    <ReactRouter.MemoryRouter>
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    </ReactRouter.MemoryRouter>
  );
};

describe('FestivalsPage 테스트', () => {
  vi.spyOn(ReactRouter, 'useParams').mockReturnValue({ areaId: '2' });
  describe('페이지 기본 구조', () => {
    test('스냅샷 테스트', () => {
      // 기본 스냅샷 테스트
      const { container } = render(
        <TestWrapper>
          <FestivalsPage />
        </TestWrapper>,
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('AI 추천 섹션', () => {
    // AI 추천 섹션 백엔드 완성 후 테스트 코드 작성 필요
    test('AI 추천 섹션이 표시된다', async () => {
      // Given: API 호출이 성공적으로 데이터를 반환할 때
      // When: FestivalsPage를 렌더링하면
      render(
        <TestWrapper>
          <FestivalsPage />
        </TestWrapper>,
      );

      // Then: AI pick 섹션 제목이 표시되어야 한다
      await waitFor(() => {
        expect(screen.getByText('AI pick')).toBeInTheDocument();
      });
    });
  });

  describe('축제 목록 섹션', () => {
    test('API에서 가져온 축제 목록이 표시된다', async () => {
      // Given: API가 축제 데이터를 성공적으로 반환할 때

      // When: FestivalsPage를 렌더링하면
      render(
        <TestWrapper>
          <FestivalsPage />
        </TestWrapper>,
      );

      // Then: Festivals 섹션 내에서 축제들이 표시되어야 한다
      await waitFor(() => {
        const festivalsSection = screen.getByText('Festivals').closest('section');
        expect(festivalsSection).toBeInTheDocument();

        // Festivals 섹션 내에서만 축제 제목들을 검색
        expect(festivalsSection).toHaveTextContent(festivalsMockData.content[0].title);
        expect(festivalsSection).toHaveTextContent(festivalsMockData.content[1].title);
        expect(festivalsSection).toHaveTextContent(festivalsMockData.content[2].title);
      });

      // 축제 정보가 올바르게 표시되어야 한다
      const festivalsSection = screen.getByText('Festivals').closest('section');
      expect(festivalsSection).toHaveTextContent(festivalsMockData.content[0].startDate);
      expect(festivalsSection).toHaveTextContent(festivalsMockData.content[1].startDate);
      expect(festivalsSection).toHaveTextContent(festivalsMockData.content[2].startDate);
      expect(festivalsSection).toHaveTextContent(
        festivalsMockData.content[0].addr1.split(' ').slice(0, 2).join(' '),
      );
      expect(festivalsSection).toHaveTextContent(
        festivalsMockData.content[1].addr1.split(' ').slice(0, 2).join(' '),
      );
      expect(festivalsSection).toHaveTextContent(
        festivalsMockData.content[2].addr1.split(' ').slice(0, 2).join(' '),
      );
    });

    test('로딩 상태가 올바르게 표시된다', () => {
      // Given: API 호출이 진행 중일 때
      // When: FestivalsPage를 렌더링하면
      render(
        <TestWrapper>
          <FestivalsPage />
        </TestWrapper>,
      );

      // Then: 스켈레톤 UI가 표시되어야 한다
      const skeletons = document.querySelectorAll('.animate-pulse');
      expect(skeletons.length).toBeGreaterThan(0);
    });

    test('API 에러 상태가 올바르게 처리된다 (ErrorBoundary가 null을 렌더링)', async () => {
      // Given: API 호출이 실패할 때
      // When: FestivalsPage를 렌더링하면
      render(
        <TestWrapper>
          <FestivalsPage />
        </TestWrapper>,
      );

      // Then: ErrorBoundary가 null을 렌더링하므로 해당 섹션 텍스트가 나타나지 않는다
      await waitFor(() => {
        expect(screen.queryByText('Festivals')).not.toBeInTheDocument();
      });
    });
  });
});
