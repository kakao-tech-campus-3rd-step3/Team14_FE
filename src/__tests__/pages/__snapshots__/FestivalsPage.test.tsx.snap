// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FestivalsPage 테스트 > 페이지 기본 구조 > 스냅샷 테스트 1`] = `
<div
  class="w-full h-full flex flex-col items-center bg-gray-100"
>
  <div
    class="min-h-screen pt-12 pb-15 max-w-[480px] w-full relative bg-white "
  >
    <div
      class="w-full mx-auto flex flex-col items-center fixed top-0 left-0 z-999 bg-gray-100"
    >
      <div
        class="w-full max-w-[480px] h-12 border-b border-gray-300 flex items-center bg-white justify-between"
      >
        <div
          class="flex-1 flex justify-start"
        >
          <button
            class="cursor-pointer flex items-center justify-center w-fit bg-transparent text-gray-900 text-md rounded-xl px-5 py-3 undefined"
            type="button"
          >
            <svg
              class="size-8"
              fill="none"
              height="28"
              stroke="black"
              stroke-width="3"
              viewBox="0 0 24 24"
              width="28"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16 4 L8 12 L16 20"
              />
            </svg>
          </button>
        </div>
        <div
          class="flex-1 flex justify-center"
        >
          <h1
            class="text-xl font-bold"
          />
        </div>
        <div
          class="flex-1 flex justify-end"
        />
      </div>
    </div>
    <div
      class="flex flex-col items-center px-8 py-4 gap-12"
    >
      <section
        class="w-full"
      >
        <div
          class="mb-6"
        >
          <div
            class="bg-gray-200 animate-pulse rounded h-8 w-32"
          />
        </div>
        <div
          class="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 gap-6"
        >
          <div
            class="bg-white rounded-xl overflow-hidden shadow-lg"
          >
            <div
              class="aspect-[3/2] w-full"
            >
              <div
                class="bg-gray-200 animate-pulse rounded w-full h-full"
              />
            </div>
            <div
              class="p-4 space-y-3"
            >
              <div
                class="space-y-2"
              >
                <div
                  class="bg-gray-200 animate-pulse rounded h-5 w-full"
                />
                <div
                  class="bg-gray-200 animate-pulse rounded h-5 w-3/4"
                />
              </div>
              <div
                class="space-y-1"
              >
                <div
                  class="bg-gray-200 animate-pulse rounded h-4 w-2/3"
                />
                <div
                  class="bg-gray-200 animate-pulse rounded h-4 w-1/2"
                />
              </div>
            </div>
          </div>
          <div
            class="bg-white rounded-xl overflow-hidden shadow-lg"
          >
            <div
              class="aspect-[3/2] w-full"
            >
              <div
                class="bg-gray-200 animate-pulse rounded w-full h-full"
              />
            </div>
            <div
              class="p-4 space-y-3"
            >
              <div
                class="space-y-2"
              >
                <div
                  class="bg-gray-200 animate-pulse rounded h-5 w-full"
                />
                <div
                  class="bg-gray-200 animate-pulse rounded h-5 w-3/4"
                />
              </div>
              <div
                class="space-y-1"
              >
                <div
                  class="bg-gray-200 animate-pulse rounded h-4 w-2/3"
                />
                <div
                  class="bg-gray-200 animate-pulse rounded h-4 w-1/2"
                />
              </div>
            </div>
          </div>
          <div
            class="bg-white rounded-xl overflow-hidden shadow-lg"
          >
            <div
              class="aspect-[3/2] w-full"
            >
              <div
                class="bg-gray-200 animate-pulse rounded w-full h-full"
              />
            </div>
            <div
              class="p-4 space-y-3"
            >
              <div
                class="space-y-2"
              >
                <div
                  class="bg-gray-200 animate-pulse rounded h-5 w-full"
                />
                <div
                  class="bg-gray-200 animate-pulse rounded h-5 w-3/4"
                />
              </div>
              <div
                class="space-y-1"
              >
                <div
                  class="bg-gray-200 animate-pulse rounded h-4 w-2/3"
                />
                <div
                  class="bg-gray-200 animate-pulse rounded h-4 w-1/2"
                />
              </div>
            </div>
          </div>
        </div>
      </section>
      <section
        class="w-full"
      >
        <div
          class="mb-6"
        >
          <div
            class="bg-gray-200 animate-pulse rounded h-8 w-32"
          />
        </div>
        <div
          class="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 gap-6"
        >
          <div
            class="bg-white rounded-xl overflow-hidden shadow-lg"
          >
            <div
              class="aspect-[3/2] w-full"
            >
              <div
                class="bg-gray-200 animate-pulse rounded w-full h-full"
              />
            </div>
            <div
              class="p-4 space-y-3"
            >
              <div
                class="space-y-2"
              >
                <div
                  class="bg-gray-200 animate-pulse rounded h-5 w-full"
                />
                <div
                  class="bg-gray-200 animate-pulse rounded h-5 w-3/4"
                />
              </div>
              <div
                class="space-y-1"
              >
                <div
                  class="bg-gray-200 animate-pulse rounded h-4 w-2/3"
                />
                <div
                  class="bg-gray-200 animate-pulse rounded h-4 w-1/2"
                />
              </div>
            </div>
          </div>
          <div
            class="bg-white rounded-xl overflow-hidden shadow-lg"
          >
            <div
              class="aspect-[3/2] w-full"
            >
              <div
                class="bg-gray-200 animate-pulse rounded w-full h-full"
              />
            </div>
            <div
              class="p-4 space-y-3"
            >
              <div
                class="space-y-2"
              >
                <div
                  class="bg-gray-200 animate-pulse rounded h-5 w-full"
                />
                <div
                  class="bg-gray-200 animate-pulse rounded h-5 w-3/4"
                />
              </div>
              <div
                class="space-y-1"
              >
                <div
                  class="bg-gray-200 animate-pulse rounded h-4 w-2/3"
                />
                <div
                  class="bg-gray-200 animate-pulse rounded h-4 w-1/2"
                />
              </div>
            </div>
          </div>
          <div
            class="bg-white rounded-xl overflow-hidden shadow-lg"
          >
            <div
              class="aspect-[3/2] w-full"
            >
              <div
                class="bg-gray-200 animate-pulse rounded w-full h-full"
              />
            </div>
            <div
              class="p-4 space-y-3"
            >
              <div
                class="space-y-2"
              >
                <div
                  class="bg-gray-200 animate-pulse rounded h-5 w-full"
                />
                <div
                  class="bg-gray-200 animate-pulse rounded h-5 w-3/4"
                />
              </div>
              <div
                class="space-y-1"
              >
                <div
                  class="bg-gray-200 animate-pulse rounded h-4 w-2/3"
                />
                <div
                  class="bg-gray-200 animate-pulse rounded h-4 w-1/2"
                />
              </div>
            </div>
          </div>
        </div>
      </section>
    </div>
    <div
      class="w-full mx-auto flex flex-col items-center fixed bottom-0 left-0 z-999"
    >
      <div
        class="w-full max-w-[480px] h-15 border-t border-gray-300 flex items-center bg-white"
      >
        <a
          class="flex-1 flex items-center justify-center"
          data-discover="true"
          href="/"
        >
          <button
            class="cursor-pointer flex items-center justify-center w-fit bg-transparent text-gray-900 text-md rounded-xl px-5 py-3 flex-col text-xs pt-2 pb-3 px-3"
            type="button"
          >
            <svg
              class="size-6"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M3.99999 10L12 3L20 10L20 20H15V16C15 15.2044 14.6839 14.4413 14.1213 13.8787C13.5587 13.3161 12.7956 13 12 13C11.2043 13 10.4413 13.3161 9.87868 13.8787C9.31607 14.4413 9 15.2043 9 16V20H4L3.99999 10Z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
            <p
              class="font-normal"
            >
              홈
            </p>
          </button>
        </a>
        <a
          class="flex-1 flex items-center justify-center"
          data-discover="true"
          href="/search"
        >
          <button
            class="cursor-pointer flex items-center justify-center w-fit bg-transparent text-gray-900 text-md rounded-xl px-5 py-3 flex-col text-xs pt-2 pb-3 px-3"
            type="button"
          >
            <svg
              class="size-6"
              fill="none"
              height="24"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15.7955 15.8111L21 21M18 10.5C18 14.6421 14.6421 18 10.5 18C6.35786 18 3 14.6421 3 10.5C3 6.35786 6.35786 3 10.5 3C14.6421 3 18 6.35786 18 10.5Z"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
            <p
              class="font-normal"
            >
              검색
            </p>
          </button>
        </a>
        <a
          class="flex-1 flex items-center justify-center"
          data-discover="true"
          href="/pick/:areaId"
        >
          <button
            class="cursor-pointer flex items-center justify-center w-fit bg-transparent text-gray-900 text-md rounded-xl px-5 py-3 flex-col text-xs pt-2 pb-3 px-3"
            type="button"
          >
            <svg
              class="size-6"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-width="4"
              viewBox="0 0 64 64"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="34.52"
                cy="11.43"
                r="5.82"
              />
              <circle
                cx="53.63"
                cy="31.6"
                r="5.82"
              />
              <circle
                cx="34.52"
                cy="50.57"
                r="5.82"
              />
              <circle
                cx="15.16"
                cy="42.03"
                r="5.82"
              />
              <circle
                cx="15.16"
                cy="19.27"
                r="5.82"
              />
              <circle
                cx="34.51"
                cy="29.27"
                r="4.7"
              />
              <line
                x1="20.17"
                x2="28.9"
                y1="16.3"
                y2="12.93"
              />
              <line
                x1="38.6"
                x2="49.48"
                y1="15.59"
                y2="27.52"
              />
              <line
                x1="50.07"
                x2="38.67"
                y1="36.2"
                y2="46.49"
              />
              <line
                x1="18.36"
                x2="30.91"
                y1="24.13"
                y2="46.01"
              />
              <line
                x1="20.31"
                x2="28.7"
                y1="44.74"
                y2="48.63"
              />
              <line
                x1="17.34"
                x2="31.37"
                y1="36.63"
                y2="16.32"
              />
              <line
                x1="20.52"
                x2="30.34"
                y1="21.55"
                y2="27.1"
              />
              <line
                x1="39.22"
                x2="47.81"
                y1="29.8"
                y2="30.45"
              />
              <line
                x1="34.51"
                x2="34.52"
                y1="33.98"
                y2="44.74"
              />
            </svg>
            <p
              class="font-normal"
            >
              AI 추천
            </p>
          </button>
        </a>
        <a
          class="flex-1 flex items-center justify-center"
          data-discover="true"
          href="/mypage"
        >
          <button
            class="cursor-pointer flex items-center justify-center w-fit bg-transparent text-gray-900 text-md rounded-xl px-5 py-3 flex-col text-xs pt-2 pb-3 px-3"
            type="button"
          >
            <svg
              class="size-6"
              fill="none"
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5 21C5 17.134 8.13401 14 12 14C15.866 14 19 17.134 19 21M16 7C16 9.20914 14.2091 11 12 11C9.79086 11 8 9.20914 8 7C8 4.79086 9.79086 3 12 3C14.2091 3 16 4.79086 16 7Z"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
            <p
              class="font-normal"
            >
              My
            </p>
          </button>
        </a>
      </div>
    </div>
  </div>
</div>
`;
